### PR TITLE
[WIP] Flash Attention pallas kernel logit softcapping

### LINKED ISSUE #23300: Flash attention soft capping support
In Jax experimental pallas kernels for TPU , there is support for attn logits softcapping for paged attention but not for flash attention.
If support can be added for pallas flash kernels as well, as it can then be used in pytorch xla as well as vllm implementation.
Gemma 2 9b model works even with logit softcapping but 27 b doesn't. 

[PR for support of soft capping for Paged Attention](https://github.com/google/jax/commit/f0e36d5083457278339387285b06f170d28d87ac)

[Pytorch xla custom kernel integration for paged attention](https://github.com/pytorch/xla/pull/7704)

[Need for flash attention support for running Gemma 2 with VLLM on TPUs](https://github.com/vllm-project/vllm/issues/7950)

### PR DESCRIPTION
**Pull request by sshkhr**
Fixes #23300 : Adds support to cap the logit values of the attention layers `logits ← soft_cap * tanh(logits/soft_cap)`

Currently, it is only implemented in the forward pass, as I wanted to get some feedback from the maintainers (since it's my first patch to JAX). The implementation is based off the official [Pytorch implementation for Gemma](https://github.com/google/gemma_pytorch/blob/014acb7ac4563a5f77c76d7ff98f31b568c16508/gemma/model.py#L324) released by Google

I had a question regarding the implementation here: in the official [FlashAttention PR #1025](https://github.com/Dao-AILab/flash-attention/pull/1025#issuecomment-2211414707) discussion, they fuse the logit softcapping and the softmax scaling operations into one as *"hopefully by multiplying these constants together beforehand we can reduce the number of instructions."* Would a similar logic apply to pallas? Or will the speedup from fusing these operations be offset by the additional control divergence (new if-else block within `if sm_scale != 1.0:`)? I am not familiar with how the JAX compiler resolves this under the hood.

TODOs:

- [ ] Add backward pass implementation

### CODE REVIEW

**Comment by sshkhr** [comment]
Hi @sharadmv any initial comments or notes (regrading fusion)? Otherwise I can continue with the backward pass

**Review by sharadmv — COMMENTED** [review]
~Were you also going to include bwd pass support?~

NVM just read your comment.
